pd-client/client.go

```diff
@@ -43,77 +50,279 @@ type Client interface {
+		if err != nil {
+			continue
+		}
+		leader, err := client.GetLeader()
+		if err != nil {
+			continue
+		}
+		c.mu.RLock()
+		changed := c.leader != leader.GetAddr()
+		c.mu.RUnlock()
+		if changed {
+			if err = c.switchLeader(leader.GetAddr()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	}
+	return errors.Errorf("failed to get leader from %v", c.urls)
+}
+
+func (c *client) switchLeader(addr string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	log.Infof("[pd] leader switches to: %v, previous: %v", addr, c.leader)
+	if _, ok := c.clients[addr]; !ok {
+		cc, err := grpc.Dial(addr, grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+			u, err := url.Parse(addr)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			return net.Dial(u.Scheme, u.Host)
+		}), grpc.WithInsecure())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.clients[addr] = pdpb2.NewPDClient(cc)
+	}
+	c.leader = addr
+	return nil
+}
+
+func (c *client) leaderLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-c.checkLeaderCh:
+		case <-time.After(time.Minute):
+		case <-c.quit:
+			return
+		}
+
+		if err := c.updateLeader(); err != nil {
+			log.Errorf("[pd] failed updateLeader: %v", err)
+		}
+	}
+}
+
+func (c *client) tsLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case first := <-c.tsoRequests:
+			c.procTSORequests(first)
+		case <-c.quit:
+			return
+		}
+	}
+}
+
+func (c *client) procTSORequests(first *tsoRequest) {
+	start := time.Now()
+	ctx, cancel := context.WithTimeout(context.Background(), pdTimeout)
+
+	pendingCount := len(c.tsoRequests)
+	resp, err := c.leaderClient().Tso(ctx, &pdpb2.TsoRequest{
+		Header: &pdpb2.RequestHeader{ClusterId: c.clusterID},
+		Count:  uint32(pendingCount + 1),
+	})
+	cancel()
+	requestDuration.WithLabelValues("tso").Observe(time.Since(start).Seconds())
+	if err == nil && resp.GetCount() != uint32(pendingCount+1) {
+		err = errors.New("[pd] tso length in rpc response is incorrect")
+	}
+	if err != nil {
+		c.finishTSORequest(first, 0, 0, errors.Trace(err))
+		for i := 0; i < pendingCount; i++ {
+			c.finishTSORequest(<-c.tsoRequests, 0, 0, errors.Trace(err))
+		}
+		return
+	}
+
+	physical, logical := resp.GetTimestamp().GetPhysical(), resp.GetTimestamp().GetLogical()
+	c.finishTSORequest(first, physical, logical, nil)
+	for i := 0; i < pendingCount; i++ {
+		logical--
+		c.finishTSORequest(<-c.tsoRequests, physical, logical, nil)
+	}
+}
+
+func (c *client) finishTSORequest(req *tsoRequest, physical, logical int64, err error) {
+	req.physical, req.logical = physical, logical
+	req.done <- err
 }
 
 func (c *client) Close() {
-	c.worker.stop(errors.New("[pd] pd-client closing"))
+	close(c.quit)
+	c.wg.Wait()
+
+	n := len(c.tsoRequests)
+	for i := 0; i < n; i++ {
+		req := <-c.tsoRequests
+		req.done <- errors.New("[pd] closing")
+	}
+}
+
+func (c *client) leaderClient() pdpb2.PDClient {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.clients[c.leader]
```

Comment:
if we have not gotten the leader, this function will return nil? TSO loop may panic? Is it a risk?